beginner_source/Pretraining_Vgg_from_scratch.py

```diff
@@ -0,0 +1,665 @@
+# We recommend using GPU for this tutorial.
+# 
+
+device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+
+######################################################################
+# Worth point of this tutorial
+# ----------------------------
+# 
+
+
+######################################################################
+# -  We train the model from scratch using only the configuration
+#    presented in the paper.
+# 
+#    -  we do not use future method, like ``Batch normalization``,Adam , He
+#       initialization.
+# 
+# -  You can apply to ImageNet Data.
+# 
+#    -  If you can download the ImageNet Data(140GB), you can apply this
+#       tutorial to reproduce Original VGG.
+# 
+# -  You can learn VGG within the training time suggested in the paper.
+# 
+
+
+######################################################################
+# Why VGG is so popular ?
+# -----------------------
+# 
+
+
+######################################################################
+# VGG became a model that attracted attention because it succeeded in
+# building deeper layers and dramatically shortening the training time
+# compared to ``alexnet``, which was the SOTA model at the time.:
+# 
+
+
+######################################################################
+# VGG Configuration
+# -----------------
+# 
+
+
+######################################################################
+# We define some configurations suggested in VGG paper . Details about
+# this configuration will be explained below section.
+# 
+
+DatasetName = 'Cifar' # CIFAR ,CIFAR10, MNIST , ImageNet
+
+## model configuration
+
+num_classes =   100
+# ``Caltech`` 257 CIFAR 100  CIFAR10 10 ,MNIST 10 ImageNet 1000
+model_version = None ## you must configure it.
+
+## data configuration
+
+train_min = 256
+train_max = None
+test_min = 256
+test_max = 256
+
+## train configuration
+
+batch_size = 32
+lr = 1e-2
+momentum = 0.9
+weight_decay  = 5e-4
+lr_factor = 0.1
+epoch = 10
+clip= None # model D grad clip 0.7
+
+
+update_count = int(256/batch_size)
+accum_step = int(256/batch_size)
+eval_step =26 * accum_step  ## ``Caltech`` 5 CIFAR 5 MNIST 6 , CIFAR10 5 ImageNet  26
+
+
+## model configuration
+xavier_count= 4
+
+last_xavier = -8  ##
+
+except_xavier = None
+
+model_layers =None
+
+
+
+######################################################################
+# | If your GPU memory is 24GB ,The maximum batch size is 128. But, if you
+#   use Colab , We recommend using 32 .
+# | You can modify the batch size according to your preference.
+# 
+
+
+######################################################################
+# Define dataset
+# --------------
+# 
+
+
+######################################################################
+# We use ``CIFAR100`` Dataset in this tutorial. In VGG paper , the authors
+# scales image ``isotropically`` . Then , they apply
+# ``Normalization``,``RandomCrop``,``HorizontalFlip`` . So , we need to override
+# CIFAR100 class to apply preprocessing.
```

Comment:
I suggest discussing the purpose of these preprocessing techniques. Normalization is utilized to bring RGB values between 0 and 1; gradient-based learning operates much better on this distribution of values in practice. Random crop and horizontal flip are data augmentation techniques with the purpose of improving generalization to data outside of the training set.